### DISCUSSION TITLE
Creating Async Engine and Calling Functions With run_sync

### DISCUSSION THREAD

**Discussion by unknown0anon** [sqlalchemy/sqlalchemy#7355] [Usage Questions] [answered]
I am new to using async/await in Python and to using SQLAlchemy, so this might be something really basic. But I am trying to understand the different error responses I am getting when calling functions with an async engine and run_sync.

I am working on code I didn't originally write, so I need to keep to certain pre-established constraints of async patterns.
Let's say I have a class that is the interface to a database. We want to instantiate this class and immediately have an engine and session to work with. 
Then I want to call specific class methods like `clear_db`, `create_db`, etc.
The 3 methods in this class all require a sync engine:  `MetaData.create_all()`), and the functions `create_db` and `drop_db` which are using the `sqlalchemy_utils` functions.

```
import asyncio
from sqlalchemy_utils import create_database, database_exists, drop_database
from sqlalchemy.ext.asyncio import AsyncSession, create_async_engine
from sqlalchemy.orm import sessionmaker
from base import Base
ENGINE = 'postgresql+asyncpg://postgres@localhost:5432/ydb'

class DbInterface:
    def __init__(self, config):
        self.engine = create_async_engine(ENGINE, echo=True, future=True)
        self.session = sessionmaker(self.engine, expire_on_commit=False, class_=AsyncSession)

    async def create_db(self):
        async with self.engine.begin() as conn:
            if not database_exists(self.engine.url):
                await conn.run_sync(create_database(self.engine.url))

    async def drop_db(self):
        async with self.engine.begin() as conn:
            await conn.run_sync(drop_database(self.engine.url))


    async def create_all(self):
        async with self.engine.begin() as conn:
            await conn.run_sync(Base.metadata.create_all)
```

I instantiate the class like this:
```
ydb = DbInterface(ENGINE)
```
And call it like:
```
asyncio.run(ydb.create_all()) # <- This works
```
When I try to call `drop_db` in the same way:
```
asyncio.run(ydb.drop_db())
```
I receive the error:
```
File "...python3.7/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 68, in await_only
    "greenlet_spawn has not been called; can't call await_() here. "
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)
sys:1: RuntimeWarning: coroutine 'connect' was never awaited
```

If I try to change the method and define it regularly like:
```
def drop_db(self):
    drop_database(ENGINE)
```
And call it regularly like:
```
ydb = DbInterface(ENGINE)
ydb.drop_db()
```

I receive error:
```
"...python3.7/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 68, in await_only
    "greenlet_spawn has not been called; can't call await_() here. "
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)
sys:1: RuntimeWarning: coroutine 'connect' was never awaited
```

I have tried a couple of other ways of defining these methods but I am receiving the same error.
Why does function `create_all` run when called with asyncio, but `create_db` does not?

**Comment by CaselIT** [comment, marked as answer]
Hi,

The issue seems to be that `sqlalchemy_utils` does not currently seem to support async engines for create/drop/exists functions. As you can see [here](https://github.com/kvesteri/sqlalchemy-utils/blob/f491aa99514560216bd79985d36651642a951737/sqlalchemy_utils/functions/database.py#L462) it creates new sync engines that will not work properly when used with an async driver.

You should probably ask in the `sqlalchemy_utils` repository if they have a current workaround. What I could think of at the moment is either:
- have also a sync dialect available, so that you can use it to run there management functions
- implement the logic of these function in your code using an async engine

**Reply by unknown0anon** [reply]
Thanks for your response!
Yes, that is the workaround I have been using - Passing an instance of a sync engine to  these functions. 
I had been trying to clean up the code by using only an async engine. 
I thought `run_sync` would work with `sqlalchemy_utils` as well, but your response clarified that for me :)

**Reply by CaselIT** [reply]
Now that you mention it, it mighty, I have not tried, but maybe you can check if running them like this works
```py
    async def create_db(self):
        async with self.engine.begin() as conn:
            if not conn.run_sync(database_exists, self.engine.url):
                await conn.run_sync(create_database, self.engine.url)

    async def drop_db(self):
        async with self.engine.begin() as conn:
            await conn.run_sync(drop_database, self.engine.url)
```

**Reply by unknown0anon** [reply]
Unfortunately these suggestions don't work, but thanks for the idea :)

(Calling `create_db` as defined above errors with:
```
 RuntimeWarning: coroutine 'AsyncConnection.run_sync' was never awaited 
 if not conn.run_sync(database_exists, self.engine.url):
 RuntimeWarning: Enable tracemalloc to get the object allocation traceback
Process finished with exit code 0
```

Calling `drop_db` as defined above errors with:
```
line 68, in await_only  "greenlet_spawn has not been called; can't call await_() here. "
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)
sys:1: RuntimeWarning: coroutine 'connect' was never awaited
```

**Comment by unknown0anon** [comment]
Hi again @CaselIT !

I am experiencing a very similar issue as already described above, so I am commenting here, but if it would be better to open a new thread, just let me know :)

I am using an Augmented Declarative Base and an async Postgres engine with the SA ORM. The tables are not held in `models.py` but are committed directly to the database by parsing a JSON file that contains all the table schemas. Because of this, I can't import the models at the top of the script ` from models import ThisTable`.

I think `automap` will work well for this.

I am setting it up in the same way that I've set up using the `async` engine with `Base.metadata.create_all` which works fine with the async engine.     

This works:
```
ENGINE = 'postgresql+asyncpg://postgres@localhost:5432/db'
engine = create_async_engine(ENGINE, echo=True, future=True)

async def create_all():
        async with engine.begin() as conn:
            # Do things to set up database
            await conn.run_sync(Base.metadata.create_all)

asyncio.run(create_all())
```

When trying to set up in the same way to use `automap` and execute a simple update like this:
```
ENGINE = 'postgresql+asyncpg://postgres@localhost:5432/db'
engine = create_async_engine(ENGINE, echo=True, future=True)

async def reflect_all():
    async with engine.begin() as conn:
        await conn.run_sync(Base.prepare(engine, reflect=True, schema='public'))

        AppUser = Base.classes.appuser

        session = Session(engine)

        session.execute(
            update(AppUser).
            where(AppUser.name == "Alice").
            values(role="admin"))

asyncio.run(reflect_all())
```

It errors with this stack trace:

```
Traceback (most recent call last):
  File "/~scratch.py", line 45, in <module>
    asyncio.run(reflect_all())
  File "/usr/lib/python3.7/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/lib/python3.7/asyncio/base_events.py", line 584, in run_until_complete
    return future.result()
  File "~scratch.py", line 14, in reflect_all
    await conn.run_sync(Base.prepare(engine, reflect=True, schema='public'))
  File "<string>", line 2, in prepare
  File "~env3/lib/python3.7/site-packages/sqlalchemy/util/deprecations.py", line 298, in warned
    return fn(*args, **kwargs)
  File "~env3/lib/python3.7/site-packages/sqlalchemy/ext/automap.py", line 872, in prepare
    cls.metadata.reflect(autoload_with, **opts)
  File "~env3/lib/python3.7/site-packages/sqlalchemy/sql/schema.py", line 4691, in reflect
    with inspection.inspect(bind)._inspection_context() as insp:
  File "~env3/lib/python3.7/site-packages/sqlalchemy/inspection.py", line 64, in inspect
    ret = reg(subject)
  File "~env3/lib/python3.7/site-packages/sqlalchemy/ext/asyncio/engine.py", line 812, in _no_insp_for_async_engine_xyet
    code="xd3s",
sqlalchemy.exc.NoInspectionAvailable: Inspection on an AsyncEngine is currently not supported. Please obtain a connection then use ``conn.run_sync`` to pass a callable where it's possible to call ``inspect`` on the passed connection. (Background on this error at: https://sqlalche.me/e/14/xd3s)
```

This says that I need to obtain a connection and run it with ``conn.run_sync``.
I am doing that with 
```   
 async with engine.begin() as conn:
        await conn.run_sync(Base.prepare(engine, reflect=True, schema='public'))
```
It's unclear to me why this error is occurring and why `conn.run_sync` works with `Base.metadata.create_all` but not with `Base.prepare`
I've verified that the automapping works well with the existing database when I pass it a `sync` engine connection instead of `async`. I suppose I could pass it an instance of a `sync` engine instead, but I would really prefer to keep the code as clean as possible without passing around multiple engine instances. 

As well, I am trying to deepen my knowledge of and proficiency with SQLAlchemy and its inner-workings, so it would be great to gain more insight into this. 
Thanks for any insight you can offer! :)

**Reply by CaselIT** [reply]
Hi,

Be careful that this is not the correct way of using run_sync
`await conn.run_sync(Base.prepare(engine, reflect=True, schema='public'))`
Here you are calling prepare and passing the result to run_sync.

The proper way of using it is to pass a callable, prepare in this case, and the arguments.

Try using
`await conn.run_sync(Base.prepare, reflect=True, schema='public')`
Engine is not needed because run_sync will pass the connection as first argument

**Reply by unknown0anon** [reply]
Thank you for the explanation!
